lib/finch/request.ex

```diff
@@ -58,9 +71,34 @@ defmodule Finch.Request do
           headers: headers(),
           body: body(),
           query: String.t() | nil,
-          unix_socket: String.t() | nil
+          unix_socket: String.t() | nil,
+          private: private_metdata() | nil
         }
 
+  @doc """
+  Sets a new **private** key and value in the request metadata. This storage is meant to be used by libraries
+  and frameworks to inject information about the request that needs to be retrieved later on, for example,
+  from handlers that consume `Finch.Telemetry` events.
+  """
+  @spec put_private(t(), key :: atom(), value :: term()) :: t()
+  def put_private(%__MODULE__{private: private} = request, key, value) do
+    if not is_atom(key) do
+      raise ArgumentError, """
+      got unsupported private metadata key #{inspect(key)}
+      Only atoms are allowed as keys of the `private` field.
+      """
+    end
+
+    updated_private =
+      if private == nil do
+        %{key => value}
+      else
+        Map.put(private, key, value)
+      end
+
+    %{request | private: updated_private}
+  end
```

Comment:
> Just a nitpick, but I would prefer to break this up a bit:

👍

> Maybe we should also consider defaulting the value of the `:private` field to an empty map, then we can remove the nil check?

Yeah that's what I was thinking but then my inner Rust programmer told me to avoid the allocation 🤣 (not sure if this even makes sense / is the case with the `%{}` literal).

Reply:
@sneako I went ahead and used a default empty map for the `:private` field, let me know what you prefer and otherwise I can revert that last commit.

Reply:
Hah, yeah I was also considering the memory implications and there is a slight difference, but probably not significant.
```
iex(1)> :erts_debug.size(nil)
0
iex(2)> :erts_debug.size(%{})
4
```

Reply:
Oh good thinking, I wonder if there are any downsides to using a keyword list instead here :thinking:

Reply:
I don't think so, the docs of `Plug.Conn` uses even the `Access` protocol syntax in the examples: `conn.private[:my_plug_hello]`, which also works with keyword lists.

The only reason to go with map is again to keep the similarity with `Plug.Conn`, where `:private` is a map :)

Reply:
Yeah, I think we should just stick with the map :)

Reply:
Agreed!